File: src/prod/src/Management/ImageModel/StoreLayoutSpecification.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"

using namespace std;
using namespace Common;
using namespace Management;
using namespace ImageModel;

GlobalWString StoreLayoutSpecification::Store = make_global<wstring>(L"Store");

StoreLayoutSpecification::StoreLayoutSpecification()
	: storeRoot_(Store)
{
}

StoreLayoutSpecification::StoreLayoutSpecification(wstring const & storeRoot)
	: storeRoot_(Path::Combine(storeRoot, *Store))
{
}

wstring StoreLayoutSpecification::GetApplicationManifestFile(
    wstring const & applicationTypeName,
    wstring const & applicationTypeVersion) const
{
    wstring applicationManifestFileName(L"ApplicationManifest.");
    applicationManifestFileName.append(applicationTypeVersion);
    applicationManifestFileName.append(L".xml");

    return Path::Combine(
        GetApplicationTypeFolder(applicationTypeName), 
        applicationManifestFileName);
}

wstring StoreLayoutSpecification::GetApplicationInstanceFile(
    wstring const & applicationTypeName,
    wstring const & applicationId,
    wstring const & applicationInstanceVersion) const
{
    wstring applicationInstanceFileName(L"ApplicationInstance.");
    applicationInstanceFileName.append(applicationInstanceVersion);
    applicationInstanceFileName.append(L".xml");

    return Path::Combine(
        GetApplicationInstanceFolder(applicationTypeName, applicationId),
        applicationInstanceFileName);
}

wstring StoreLayoutSpecification::GetApplicationPackageFile(
    wstring const & applicationTypeName,
    wstring const & applicationId,
    wstring const & applicationRolloutVersion) const
{
    wstring applicationInstanceFileName(L"ApplicationPackage.");
    applicationInstanceFileName.append(applicationRolloutVersion);
    applicationInstanceFileName.append(L".xml");

    return Path::Combine(
        GetApplicationInstanceFolder(applicationTypeName, applicationId),
        applicationInstanceFileName);
}

wstring StoreLayoutSpecification::GetServicePackageFile(
    wstring const & applicationTypeName,
    wstring const & applicationId,
    wstring const & servicePackageName,
    wstring const & servicePackageRolloutVersion) const
{
    return Path::Combine(
        GetApplicationInstanceFolder(applicationTypeName, applicationId),
        FileNamePattern::GetServicePackageFileName(servicePackageName, servicePackageRolloutVersion));
}

wstring StoreLayoutSpecification::GetServiceManifestFile(
    wstring const & applicationTypeName,
    wstring const & serviceManifestName,
    wstring const & serviceManifestVersion) const
{
    return Path::Combine(
        GetApplicationTypeFolder(applicationTypeName),
        FileNamePattern::GetServiceManifestFileName(serviceManifestName, serviceManifestVersion));
}

wstring StoreLayoutSpecification::GetServiceManifestChecksumFile(
    wstring const & applicationTypeName,
    wstring const & serviceManifestName,
    wstring const & serviceManifestVersion) const
{
    return FileNamePattern::GetChecksumFileName(GetServiceManifestFile(
        applicationTypeName,
        serviceManifestName,
        serviceManifestVersion));
}

wstring StoreLayoutSpecification::GetCodePackageFolder(
    wstring const & applicationTypeName,
    wstring const & serviceManifestName,
    wstring const & codePackageName,
    wstring const & codePackageVersion) const
{
    return Path::Combine(
        GetApplicationTypeFolder(applicationTypeName),
        FolderNamePattern::GetCodePackageFolderName(
            serviceManifestName,
            codePackageName,
            codePackageVersion));
}

wstring StoreLayoutSpecification::GetCodePackageChecksumFile(
    wstring const & applicationTypeName,
    wstring const & serviceManifestName,
    wstring const & codePackageName,
    wstring const & codePackageVersion) const
{
    return FileNamePattern::GetChecksumFileName(GetCodePackageFolder(
        applicationTypeName,
        serviceManifestName,
        codePackageName,
        codePackageVersion));
}

wstring StoreLayoutSpecification::GetConfigPackageFolder(
    wstring const & applicationTypeName,
    wstring const & serviceManifestName,
    wstring const & configPackageName,
    wstring const & configPackageVersion) const
{
    return Path::Combine(
        GetApplicationTypeFolder(applicationTypeName),
        FolderNamePattern::GetConfigPackageFolderName(
            serviceManifestName,
            configPackageName,
            configPackageVersion));
}

wstring StoreLayoutSpecification::GetConfigPackageChecksumFile(
    wstring const & applicationTypeName,
    wstring const & serviceManifestName,
    wstring const & configPackageName,
    wstring const & configPackageVersion) const
{
    return FileNamePattern::GetChecksumFileName(GetConfigPackageFolder(
        applicationTypeName,
        serviceManifestName,
        configPackageName,
        configPackageVersion));
}

wstring StoreLayoutSpecification::GetDataPackageFolder(
    wstring const & applicationTypeName,
    wstring const & serviceManifestName,
    wstring const & dataPackageName,
    wstring const & dataPackageVersion) const
{
    return Path::Combine(
        GetApplicationTypeFolder(applicationTypeName),
        FolderNamePattern::GetDataPackageFolderName(
            serviceManifestName,
            dataPackageName,
            dataPackageVersion));
}

wstring StoreLayoutSpecification::GetDataPackageChecksumFile(
    wstring const & applicationTypeName,
    wstring const & serviceManifestName,
    wstring const & dataPackageName,
    wstring const & dataPackageVersion) const
{
    return FileNamePattern::GetChecksumFileName(GetDataPackageFolder(
        applicationTypeName,
        serviceManifestName,
        dataPackageName,
        dataPackageVersion));
}

wstring StoreLayoutSpecification::GetSettingsFile(
    wstring const & configPackageFolder) const
{
    return Path::Combine(configPackageFolder, *(FileNamePattern::SettingsFileName));
}

wstring StoreLayoutSpecification::GetApplicationTypeFolder(
    wstring const & applicationTypeName) const
{
    return Path::Combine(storeRoot_, applicationTypeName);
}

wstring StoreLayoutSpecification::GetApplicationInstanceFolder(
    wstring const & applicationTypeName, 
    wstring const & applicationId) const
{
    wstring applicationFolder = Path::Combine(GetApplicationTypeFolder(applicationTypeName), L"apps");
    return Path::Combine(applicationFolder, applicationId);
}

wstring StoreLayoutSpecification::GetSubPackageArchiveFile(
    wstring const & packageFolder) const
{
    return ImageModelUtility::GetSubPackageArchiveFileName(packageFolder);
}

wstring StoreLayoutSpecification::GetSubPackageExtractedFolder(
    wstring const & archiveFile) const
{
    return ImageModelUtility::GetSubPackageExtractedFolderName(archiveFile);
}

File: src/prod/src/managed/Api/src/System/Fabric/Common/ImageModel/StoreLayoutSpecification.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace System.Fabric.Common.ImageModel
{
    using System;
    using System.Fabric.Interop;
    using System.Reflection;

    internal class StoreLayoutSpecification
    {
        private NativeImageModel.IFabricStoreLayoutSpecification2 nativeStoreLayout;

        private StoreLayoutSpecification(NativeImageModel.IFabricStoreLayoutSpecification2 nativeStoreLayout)
        {
            this.nativeStoreLayout = nativeStoreLayout;
        }

        public static StoreLayoutSpecification Create()
        {
            return Utility.WrapNativeSyncInvokeInMTA<StoreLayoutSpecification>(StoreLayoutSpecification.CreateHelper, "StoreLayoutSpecification.Create");
        }

        public void SetRoot(string value)
        {
            Utility.WrapNativeSyncInvokeInMTA(() => this.SetRootHelper(value), "StoreLayoutSpecification.SetRoot");
        }

        public string GetRoot()
        {
            return Utility.WrapNativeSyncInvokeInMTA<string>(this.GetRootHelper, "StoreLayoutSpecification.GetRoot");
        }

        public string GetApplicationManifestFile(
            string applicationTypeName,
            string applicationTypeVersion)
        {
            return Utility.WrapNativeSyncInvokeInMTA(() => this.GetApplicationManifestFileHelper(applicationTypeName, applicationTypeVersion), "StoreLayoutSpecification.GetApplicationManifestFile");
        }

        public string GetApplicationInstanceFile(
            string applicationTypeName,
            string applicationId,
            string applicationInstanceVersion)
        {
            return Utility.WrapNativeSyncInvokeInMTA(() => this.GetApplicationInstanceFileHelper(applicationTypeName, applicationId, applicationInstanceVersion), "StoreLayoutSpecification.GetAPplicationInstanceFile");
        }

        public string GetApplicationPackageFile(
            string applicationTypeName,
            string applicationId,
            string applicationRolloutVersion)
        {
            return Utility.WrapNativeSyncInvokeInMTA(() => this.GetApplicationPackageFileHelper(applicationTypeName, applicationId, applicationRolloutVersion), "StoreLayoutSpecification.GetApplicationPackageFile");
        }

        public string GetServicePackageFile(
            string applicationTypeName,
            string applicationId,
            string servicePackageName,
            string servicePackageRolloutVersion)
        {
            return Utility.WrapNativeSyncInvokeInMTA(() => this.GetServicePackageFileHelper(applicationTypeName, applicationId, servicePackageName, servicePackageRolloutVersion), "StoreLayoutSpecification.GetServicePackageFileHelper");
        }

        public string GetServiceManifestFile(
            string applicationTypeName,
            string serviceManifestName,
            string serviceManifestVersion)
        {
            return Utility.WrapNativeSyncInvokeInMTA(() => this.GetServiceManifestFileHelper(applicationTypeName, serviceManifestName, serviceManifestVersion), "StoreLayoutSpecification.GetServiceManifestFile");
        }

        public string GetServiceManifestChecksumFile(
            string applicationTypeName,
            string serviceManifestName,
            string serviceManifestVersion)
        {
            return Utility.WrapNativeSyncInvokeInMTA(() => this.GetServiceManifestChecksumFileHelper(applicationTypeName, serviceManifestName, serviceManifestVersion), "StoreLayoutSpecification.GetServiceManifestChecksumFile");
        }

        public string GetCodePackageFolder(string applicationTypeName, string servicePackageName, string codePackageName, string codePackageVersion)
        {
            return Utility.WrapNativeSyncInvokeInMTA(() => this.GetCodePackageFolderHelper(applicationTypeName, servicePackageName, codePackageName, codePackageVersion), "StoreLayoutSpecification.GetCodePackageFolder");
        }

        public string GetCodePackageChecksumFile(string applicationTypeName, string servicePackageName, string codePackageName, string codePackageVersion)
        {
            return Utility.WrapNativeSyncInvokeInMTA(() => this.GetCodePackageChecksumFileHelper(applicationTypeName, servicePackageName, codePackageName, codePackageVersion), "StoreLayoutSpecification.GetCodePackageChecksumFile");
        }

        public string GetConfigPackageFolder(string applicationTypeName, string servicePackageName, string configPackageName, string configPackageVersion)
        {
            return Utility.WrapNativeSyncInvokeInMTA(() => this.GetConfigPackageFolderHelper(applicationTypeName, servicePackageName, configPackageName, configPackageVersion), "StoreLayoutSpecification.GetConfigPackageFolder");
        }

        public string GetConfigPackageChecksumFile(string applicationTypeName, string servicePackageName, string configPackageName, string configPackageVersion)
        {
            return Utility.WrapNativeSyncInvokeInMTA(() => this.GetConfigPackageChecksumFileHelper(applicationTypeName, servicePackageName, configPackageName, configPackageVersion), "StoreLayoutSpecification.GetConfigPackageChecksumFile");
        }

        public string GetDataPackageFolder(string applicationTypeName, string servicePackageName, string dataPackageName, string dataPackageVersion)
        {
            return Utility.WrapNativeSyncInvokeInMTA(() => this.GetDataPackageFolderHelper(applicationTypeName, servicePackageName, dataPackageName, dataPackageVersion), "StoreLayoutSpecification.GetDataPackageFolder");
        }

        public string GetDataPackageChecksumFile(string applicationTypeName, string servicePackageName, string dataPackageName, string dataPackageVersion)
        {
            return Utility.WrapNativeSyncInvokeInMTA(() => this.GetDataPackageChecksumFileHelper(applicationTypeName, servicePackageName, dataPackageName, dataPackageVersion), "StoreLayoutSpecification.GetDataPackageChecksumFile");
        }

        public string GetSettingsFile(string configPackageFolder)
        {
            return Utility.WrapNativeSyncInvokeInMTA(() => this.GetSettingsFileHelper(configPackageFolder), "StoreLayoutSpecification.GetSettingsFile");
        }

        public string GetSubPackageArchiveFile(string packageFolder)
        {
            return Utility.WrapNativeSyncInvokeInMTA(() => this.GetSubPackageArchiveFileHelper(packageFolder), "StoreLayoutSpecification.GetSubPackageArchiveFile");
        }

        #region Interop code

        private static StoreLayoutSpecification CreateHelper()
        {
            Guid iid = typeof(NativeImageModel.IFabricStoreLayoutSpecification2).GetTypeInfo().GUID;
            NativeImageModel.IFabricStoreLayoutSpecification2 newNativeStoreLayout = null;

            newNativeStoreLayout = (NativeImageModel.IFabricStoreLayoutSpecification2)NativeImageModel.FabricCreateStoreLayoutSpecification(ref iid);

            return new StoreLayoutSpecification(newNativeStoreLayout);
        }

        private void SetRootHelper(string value)
        {
            using (var pin = new PinBlittable(value))
            {
                this.nativeStoreLayout.SetRoot(pin.AddrOfPinnedObject());
            }
        }

        private string GetRootHelper()
        {
            return StringResult.FromNative(this.nativeStoreLayout.GetRoot());
        }

        private string GetApplicationManifestFileHelper(
            string applicationTypeName,
            string applicationTypeVersion)
        {
            using (var pin = new PinCollection())
            {
                var result = this.nativeStoreLayout.GetApplicationManifestFile(
                    pin.AddBlittable(applicationTypeName),
                    pin.AddBlittable(applicationTypeVersion));

                return StringResult.FromNative(result);
            }
        }

        private string GetApplicationInstanceFileHelper(
            string applicationTypeName,
            string applicationId,
            string applicationInstanceVersion)
        {
            using (var pin = new PinCollection())
            {
                var result = this.nativeStoreLayout.GetApplicationInstanceFile(
                    pin.AddBlittable(applicationTypeName),
                    pin.AddBlittable(applicationId),
                    pin.AddBlittable(applicationInstanceVersion));

                return StringResult.FromNative(result);
            }
        }

        private string GetApplicationPackageFileHelper(
            string applicationTypeName,
            string applicationId,
            string applicationRolloutVersion)
        {
            using (var pin = new PinCollection())
            {
                var result = this.nativeStoreLayout.GetApplicationPackageFile(
                    pin.AddBlittable(applicationTypeName),
                    pin.AddBlittable(applicationId),
                    pin.AddBlittable(applicationRolloutVersion));

                return StringResult.FromNative(result);
            }
        }

        private string GetServicePackageFileHelper(
            string applicationTypeName,
            string applicationId,
            string servicePackageName,
            string servicePackageRolloutVersion)
        {
            using (var pin = new PinCollection())
            {
                var result = this.nativeStoreLayout.GetServicePackageFile(
                    pin.AddBlittable(applicationTypeName),
                    pin.AddBlittable(applicationId),
                    pin.AddBlittable(servicePackageName),
                    pin.AddBlittable(servicePackageRolloutVersion));

                return StringResult.FromNative(result);
            }
        }

        private string GetServiceManifestFileHelper(
            string applicationTypeName,
            string serviceManifestName,
            string serviceManifestVersion)
        {
            using (var pin = new PinCollection())
            {
                var result = this.nativeStoreLayout.GetServiceManifestFile(
                    pin.AddBlittable(applicationTypeName),
                    pin.AddBlittable(serviceManifestName),
                    pin.AddBlittable(serviceManifestVersion));

                return StringResult.FromNative(result);
            }
        }

        private string GetServiceManifestChecksumFileHelper(
            string applicationTypeName,
            string serviceManifestName,
            string serviceManifestVersion)
        {
            using (var pin = new PinCollection())
            {
                var result = this.nativeStoreLayout.GetServiceManifestChecksumFile(
                    pin.AddBlittable(applicationTypeName),
                    pin.AddBlittable(serviceManifestName),
                    pin.AddBlittable(serviceManifestVersion));

                return StringResult.FromNative(result);
            }
        }

        private string GetCodePackageFolderHelper(
            string applicationTypeName,
            string serviceManifestName,
            string codePackageName,
            string codePackageVersion)
        {
            using (var pin = new PinCollection())
            {
                var result = this.nativeStoreLayout.GetCodePackageFolder(
                    pin.AddBlittable(applicationTypeName),
                    pin.AddBlittable(serviceManifestName),
                    pin.AddBlittable(codePackageName),
                    pin.AddBlittable(codePackageVersion));

                return StringResult.FromNative(result);
            }
        }

        private string GetCodePackageChecksumFileHelper(
            string applicationTypeName,
            string serviceManifestName,
            string codePackageName,
            string codePackageVersion)
        {
            using (var pin = new PinCollection())
            {
                var result = this.nativeStoreLayout.GetCodePackageChecksumFile(
                    pin.AddBlittable(applicationTypeName),
                    pin.AddBlittable(serviceManifestName),
                    pin.AddBlittable(codePackageName),
                    pin.AddBlittable(codePackageVersion));

                return StringResult.FromNative(result);
            }
        }

        private string GetConfigPackageFolderHelper(
            string applicationTypeName,
            string serviceManifestName,
            string configPackageName,
            string configPackageVersion)
        {
            using (var pin = new PinCollection())
            {
                var result = this.nativeStoreLayout.GetConfigPackageFolder(
                    pin.AddBlittable(applicationTypeName),
                    pin.AddBlittable(serviceManifestName),
                    pin.AddBlittable(configPackageName),
                    pin.AddBlittable(configPackageVersion));

                return StringResult.FromNative(result);
            }
        }

        private string GetConfigPackageChecksumFileHelper(
            string applicationTypeName,
            string serviceManifestName,
            string configPackageName,
            string configPackageVersion)
        {
            using (var pin = new PinCollection())
            {
                var result = this.nativeStoreLayout.GetConfigPackageChecksumFile(
                    pin.AddBlittable(applicationTypeName),
                    pin.AddBlittable(serviceManifestName),
                    pin.AddBlittable(configPackageName),
                    pin.AddBlittable(configPackageVersion));

                return StringResult.FromNative(result);
            }
        }

        private string GetDataPackageFolderHelper(
            string applicationTypeName,
            string serviceManifestName,
            string dataPackageName,
            string dataPackageVersion)
        {
            using (var pin = new PinCollection())
            {
                var result = this.nativeStoreLayout.GetDataPackageFolder(
                    pin.AddBlittable(applicationTypeName),
                    pin.AddBlittable(serviceManifestName),
                    pin.AddBlittable(dataPackageName),
                    pin.AddBlittable(dataPackageVersion));

                return StringResult.FromNative(result);
            }
        }

        private string GetDataPackageChecksumFileHelper(
            string applicationTypeName,
            string serviceManifestName,
            string dataPackageName,
            string dataPackageVersion)
        {
            using (var pin = new PinCollection())
            {
                var result = this.nativeStoreLayout.GetDataPackageChecksumFile(
                    pin.AddBlittable(applicationTypeName),
                    pin.AddBlittable(serviceManifestName),
                    pin.AddBlittable(dataPackageName),
                    pin.AddBlittable(dataPackageVersion));

                return StringResult.FromNative(result);
            }
        }

        private string GetSettingsFileHelper(string configPackageFolder)
        {
            using (var pin = new PinBlittable(configPackageFolder))
            {
                return StringResult.FromNative(
                    this.nativeStoreLayout.GetSettingsFile(pin.AddrOfPinnedObject()));
            }
        }

        private string GetSubPackageArchiveFileHelper(
            string packageFolder)
        {
            using (var pin = new PinCollection())
            {
                var result = this.nativeStoreLayout.GetSubPackageArchiveFile(
                    pin.AddBlittable(packageFolder));

                return StringResult.FromNative(result);
            }
        }

        #endregion
    }
}